marlowe-runtime/sync/Language/Marlowe/Runtime/Sync/Database/PostgreSQL/GetContractState.hs

```diff
@@ -0,0 +1,292 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+module Language.Marlowe.Runtime.Sync.Database.PostgreSQL.GetContractState
+  where
+
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.Maybe (MaybeT(MaybeT), runMaybeT)
+import Data.Binary (Binary, get)
+import Data.Binary.Get (runGet)
+import Data.ByteString (ByteString)
+import Data.ByteString.Lazy (fromStrict)
+import Data.Int (Int16, Int64)
+import qualified Data.Map as Map
+import Data.Maybe (fromJust)
+import Data.Vector (Vector)
+import qualified Data.Vector as V
+import Hasql.TH (maybeStatement, vectorStatement)
+import qualified Hasql.Transaction as T
+import Language.Marlowe (MarloweData(..), MarloweParams(MarloweParams))
+import Language.Marlowe.Protocol.Query.Types
+import Language.Marlowe.Runtime.ChainSync.Api
+  ( Address(..)
+  , AssetId(AssetId)
+  , Assets(..)
+  , BlockHeader(..)
+  , BlockHeaderHash(..)
+  , PolicyId(..)
+  , TokenName(TokenName)
+  , Tokens(Tokens)
+  , TransactionMetadata(..)
+  , TxId(..)
+  , TxOutRef(..)
+  , fromDatum
+  )
+import Language.Marlowe.Runtime.Core.Api
+  (ContractId(..), MarloweVersion(..), MarloweVersionTag(..), Payout(..), TransactionScriptOutput(..))
+import qualified Plutus.V2.Ledger.Api as PV2
+import Prelude hiding (init)
+
+getContractState :: ContractId -> T.Transaction (Maybe SomeContractState)
+getContractState (ContractId TxOutRef{..}) = runMaybeT do
+  let params = (unTxId txId, fromIntegral txIx)
+  (contractId, roleTokenMintingPolicyId, metadata, initialBlock, initialOutput) <- fmap decodeCreateResults
+    $ MaybeT
+    $ T.statement params [maybeStatement|
+      SELECT
+        createTxOut.txId :: bytea,
+        createTxOut.txIx :: smallint,
+        (ARRAY_AGG(createTxOut.metadata))[1] :: bytea?,
+        (ARRAY_AGG(createTxOut.slotNo))[1] :: bigint,
+        (ARRAY_AGG(createTxOut.blockId))[1] :: bytea,
+        (ARRAY_AGG(createTxOut.blockNo))[1] :: bigint,
+        (ARRAY_AGG(contractTxOut.rolesCurrency))[1] :: bytea,
+        (ARRAY_AGG(contractTxOut.state))[1] :: bytea,
+        (ARRAY_AGG(contractTxOut.contract))[1] :: bytea,
+        (ARRAY_AGG(txOut.address))[1] :: bytea,
+        (ARRAY_AGG(txOut.lovelace))[1] :: bigint,
+        ARRAY_REMOVE(ARRAY_AGG(txOutAsset.policyId), NULL) :: bytea[],
+        ARRAY_REMOVE(ARRAY_AGG(txOutAsset.name), NULL) :: bytea[],
+        ARRAY_REMOVE(ARRAY_AGG(txOutAsset.quantity), NULL) :: bigint[]
+      FROM marlowe.createTxOut
+      JOIN marlowe.contractTxOut USING (txId, txIx)
+      JOIN marlowe.txOut USING (txId, txIx)
+      LEFT JOIN marlowe.txOutAsset USING (txId, txIx)
+      WHERE createTxOut.txId = $1 :: bytea
+        AND createTxOut.txIx = $2 :: smallint
+      GROUP BY createTxOut.txId, createTxOut.txIx
+    |]
+  (latestBlock, latestOutput) <- lift $ maybe (initialBlock, Just initialOutput) decodeLatestResults <$> T.statement params [maybeStatement|
```

Comment:
So latest = initial for the creation tranaction, which is what we want.

Reply:
Exactly